README.md
# ISO11784Tools

### Installation

`remotes::install_github("graemediack/ISO11784Tools",build_vignettes = TRUE)`

`vignette("ISO11784Tools")`


## Introduction
ISO11784 and ISO11785 are standards describing how information should be encoded within passive integrated transponder tags (PIT tags).

ISO11784 describes the binary construction of the information stored in the tag, and ISO11785 describes how it should be read by PIT tag readers.

Despite this, there are some differences in how readers will display the id information depending on user settings and manufacturer choices.

#### Further Reading

- [WIKIPEDIA](https://en.wikipedia.org/wiki/ISO_11784_and_ISO_11785)

- [RFID NEWS](https://www.rfidnews.com/ISOstandard/ISOstandard.html)

- [ISO ORG](https://www.iso.org/standard/25881.html)

- [PRIORITY DESIGN](https://www.priority1design.com.au/fdx-b_animal_identification_protocol.html)

### ISO11784 In Practice

I encountered 4 different formats of ISO11784 ID codes whilst researching for a PIT tag database. This package is intended to facilitate transformation to and from all of these formats.

It could potentially have other functionality with respect to ISO11784, for example it could connect to ICAR and identify manufacturer of tags along with other metadata.

The 4 formats identified are:

- `Dot Hexadecimal` - e.g. **3E7.1CBE991A14**. Manufacturer Code = **3E7** (Note - this is ID is reserved for testing), Animal ID = **1CBE991A14**
- `Full Decimal` - e.g. **999123456789012**. Manufacturer Code = **999** (Note - this is ID is reserved for testing), Animal ID = **123456789012**
- `Full 64 bit Hexadecimal` in two 'flavours' that depend on whether the device reverses the transmitted binary ID before converting it to hex or not:
  - `Left Hand Flavour` - e.g. **8000F9DCBE991A14**. 
  - `Right Hand Flavour` - e.g. **2858997D3B9F0001**.

Note, all 4 examples above are the same code.

## Functions

I have created a set of functions to make converting between these formats easy.

### get_iso11784_format

pass a vector of strings containing your tag codes into this function to obtain a vector of the formats
```{r, include=T,warning = FALSE}
ISO11784Tools::get_iso11784_format(c('3E7.1CBE991A14','999123456789012','8000F9DCBE991A14','2858997D3B9F0001','blahblah'))
```

### convert_to_X Family

The convert_to_X family of functions takes a vector of strings with your tag codes and converts them into the target format
```{r, include=T,warning=F}
ISO11784Tools::convert_to_isodecimal(c('3E7.1CBE991A14','999123456789012','8000F9DCBE991A14','2858997D3B9F0001','blahblah'))
ISO11784Tools::convert_to_isodothex(c('3E7.1CBE991A14','999123456789012','8000F9DCBE991A14','2858997D3B9F0001','blahblah'))
ISO11784Tools::convert_to_iso64bitl(c('3E7.1CBE991A14','999123456789012','8000F9DCBE991A14','2858997D3B9F0001','blahblah'))
ISO11784Tools::convert_to_iso64bitr(c('3E7.1CBE991A14','999123456789012','8000F9DCBE991A14','2858997D3B9F0001','blahblah'))
```

### convert_to_all
```{r, include=T,warning=F}
ISO11784Tools::convert_to_all(c('3E7.1CBE991A14','999123456789012','8000F9DCBE991A14','2858997D3B9F0001','blahblah'))
```

`convert_to_all` takes a vector of strings and returns a 6 column tibble with detected format and all the possible conversions. It simply wraps the previous 5 functions into one and outputs as a tibble to save some time for the user

### _to_ Family

The `convert_to_X` family of functions are wrappers for lower level functions that convert to and from the `Full Decimal` format. The `Full Decimal` format and the `Dot Hexadecimal` format were the first I encountered, and seem to be the most prevalent. I chose `Full Decimal` as the primary format for this package, and created functions to convert from and to that format in the early stages of development. Hence, these functions center on `Full Decimal`.

Rather than give full examples of these I will simply list them here for reference. Each of these functions will accept a single string or a vector of strings and will return `warning` and `NA` if the format is not as expected, but will still convert anything within a vector that is the correct format.

- From `Full Decimal`:

  - `isodecimal_to_iso64bitleft`
  - `isodecimal_to_iso64bitright`
  - `isodecimal_to_isodothex`
  
- To `Full Decimal`:

  - `iso64bitleft_to_isodecimal`
  - `iso64bitright_to_isodecimal`
  - `isodothex_to_isodecimal`
  
### Lowest Level Family

The above functions are built on base converters that are included as imported functions for convenience rather than intended for direct use.

These functions are written to allow conversion between the three bases; binary, decimal, and hexadecimal, to the maximum limit of the R `numeric` data type; `0b1000000000000000000000000000000000000000000000000000000`, `18014398509481984`, and `0x40000000000000` respectively. 

`R` has limitations with regard to maximum values held in `integer` and `numeric` data types. I don't know why this is so I won't try to expand on the reason here. ISO 11784 ID codes are generally larger than the `integer` maximum, but smaller than the `numeric` maximum. Other packages might have quicker versions of these functions that I haven't found (happy to receive suggestions!), so I needed to build them from scratch.

- `binary_to_decimal`
- `binary_to_hexadecimal`
- `decimal_to_binary`
- `decimal_to_hexadecimal`
- `hexadecimal_to_binary`
- `hexadecimal_to_decimal`

### The End. Enjoy!


NEWS.md
# ISO11784Tools News

### v1.1.4 2023-02-27

removed dependency on magrittr pipe operator as it was only really used superficially

### v1.1.3 2022-09-10

fix bug character/numeric mismatch by adding more `as.character` conversions to user inputs
added NEWS.md

### v1.1.2

tidy up of documentation

### v1.1.1

removed unnecessary warning messages

## v1.1.0

add new `convert_to_all` function

### v1.0.6

Changes to documentation only

### v1.0.5

final roll out of vectorised calculations, improved speed of `from_` and `to_` family of functions

### v1.0.4

bug fix - calculation error in `binary_to_hexadecimal`

### v1.0.3

### v1.0.2

### v1.0.1

## v1.0.0










 









tests/testthat.R
library(testthat)
library(ISO11784Tools)

test_check("ISO11784Tools")


tests/testthat/test-utilities.R
#' Set up testing dataframe

df <- data.frame(isodecimal = c(NA,'999012345678901','999012345678902','999012345678903',NA,'999012345678904','999012345678905','999012345678906','999012345678907','999012345678908','999012345678909','999012345678910',NA),
  isodothex = c(NA,'3E7.02DFDC1C35','3E7.02DFDC1C36','3E7.02DFDC1C37',NA,'3E7.02DFDC1C38','3E7.02DFDC1C39','3E7.02DFDC1C3A','3E7.02DFDC1C3B','3E7.02DFDC1C3C','3E7.02DFDC1C3D','3E7.02DFDC1C3E',NA),
  iso64bitleft = c(NA,'8000F9C2DFDC1C35','8000F9C2DFDC1C36','8000F9C2DFDC1C37',NA,'8000F9C2DFDC1C38','8000F9C2DFDC1C39','8000F9C2DFDC1C3A','8000F9C2DFDC1C3B','8000F9C2DFDC1C3C','8000F9C2DFDC1C3D','8000F9C2DFDC1C3E',NA),
  iso64bitright = c(NA,'AC383BFB439F0001','6C383BFB439F0001','EC383BFB439F0001',NA,'1C383BFB439F0001','9C383BFB439F0001','5C383BFB439F0001','DC383BFB439F0001','3C383BFB439F0001','BC383BFB439F0001','7C383BFB439F0001',NA),
  stringsAsFactors = F)

test_that("get_iso11784_format Works",{
  suppressWarnings({
    expect_equal(
      get_iso11784_format(
        c(df$isodecimal[1:2],df$isodothex[1:2],df$iso64bitleft[1:2],df$iso64bitright[1:2],c('1234567890123','ABD.4567890123'))
      ),
      c('unknown','isodecimal','unknown','isodothex','unknown','iso64bitl','unknown','iso64bitr','unknown','unknown')
    )
  })
})

test_that("Convert to ISO Decimal Works",{
  suppressWarnings({
    expect_equal(convert_to_isodecimal(df$isodecimal), df$isodecimal)
    expect_equal(convert_to_isodecimal(df$isodothex), df$isodecimal)
    expect_equal(convert_to_isodecimal(df$iso64bitleft), df$isodecimal)
    expect_equal(convert_to_isodecimal(df$iso64bitright), df$isodecimal)
  })
})

test_that("Convert to ISO Dot Hexadecimal Works",{
  suppressWarnings({
  expect_equal(convert_to_isodothex(df$isodecimal), df$isodothex)
  expect_equal(convert_to_isodothex(df$isodothex), df$isodothex)
  expect_equal(convert_to_isodothex(df$iso64bitleft), df$isodothex)
  expect_equal(convert_to_isodothex(df$iso64bitright), df$isodothex)
  })
})

test_that("Convert to ISO Raw Hexadecimal (Animal bit left) Works",{
  suppressWarnings({
  expect_equal(convert_to_iso64bitl(df$isodecimal), df$iso64bitleft)
  expect_equal(convert_to_iso64bitl(df$isodothex), df$iso64bitleft)
  expect_equal(convert_to_iso64bitl(df$iso64bitleft), df$iso64bitleft)
  expect_equal(convert_to_iso64bitl(df$iso64bitright), df$iso64bitleft)
  })
})

test_that("Convert to ISO Raw Hexadecimal (Animal bit right) Works",{
  suppressWarnings({
  expect_equal(convert_to_iso64bitr(df$isodecimal), df$iso64bitright)
  expect_equal(convert_to_iso64bitr(df$isodothex), df$iso64bitright)
  expect_equal(convert_to_iso64bitr(df$iso64bitleft), df$iso64bitright)
  expect_equal(convert_to_iso64bitr(df$iso64bitright), df$iso64bitright)
  })
})


tests/testthat/test-base_converters.R
test_that("Hexadecimal to Decimal converter works",{
  expect_equal(hexadecimal_to_decimal('2a'),'42')})

test_that("Binary to Decimal converter works",{
  expect_equal(binary_to_decimal('00101010'),'42')})

test_that("Decimal to Binary converter works",{
  expect_equal(decimal_to_binary('42'),'00101010')})

test_that("Decimal to Hexadecimal converter works",{
  expect_equal(decimal_to_hexadecimal('42'),'2a')})

test_that("Hexadecimal to Binary converter works",{
  expect_equal(hexadecimal_to_binary('2a'),'00101010')})

test_that("Binary to Hexadecimal converter works",{
  expect_equal(binary_to_hexadecimal('00101010'),'2a')})

# Test extremes
# case 0
testthat::test_that("Hexadecimal to Decimal converter works",{
  suppressWarnings({
  testthat::expect_equal(hexadecimal_to_decimal('0'),'0')})
})

testthat::test_that("Binary to Decimal converter works",{
  suppressWarnings({
  testthat::expect_equal(binary_to_decimal('0'),'0')})
})

testthat::test_that("Decimal to Binary converter works",{
  suppressWarnings({
  testthat::expect_equal(decimal_to_binary('0'),'0000')})
})

testthat::test_that("Decimal to Hexadecimal converter works",{
  suppressWarnings({
  testthat::expect_equal(decimal_to_hexadecimal('0'),'0')})
})

testthat::test_that("Hexadecimal to Binary converter works",{
  suppressWarnings({
  testthat::expect_equal(hexadecimal_to_binary('0'),'0000')})
})

testthat::test_that("Binary to Hexadecimal converter works",{
  suppressWarnings({
  testthat::expect_equal(binary_to_hexadecimal('0'),'0')})
})

# Case maximum
test_that("Hexadecimal to Decimal converter works",{
  expect_equal(hexadecimal_to_decimal('40000000000000'),'18014398509481984')})

test_that("Binary to Decimal converter works",{
  expect_equal(binary_to_decimal('01000000000000000000000000000000000000000000000000000000'),'18014398509481984')})

test_that("Decimal to Binary converter works",{
  expect_equal(decimal_to_binary('18014398509481984'),'01000000000000000000000000000000000000000000000000000000')})

test_that("Decimal to Hexadecimal converter works",{
  expect_equal(decimal_to_hexadecimal('18014398509481984'),'40000000000000')})

test_that("Hexadecimal to Binary converter works",{
  expect_equal(hexadecimal_to_binary('40000000000000'),'01000000000000000000000000000000000000000000000000000000')})

test_that("Binary to Hexadecimal converter works",{
  expect_equal(binary_to_hexadecimal('01000000000000000000000000000000000000000000000000000000'),'40000000000000')})


R/base_converters.R
#' Hexadecimal to Decimal converter
#' @param hex A single hexadecimal number as a string. Max '40000000000000'
#' @return A single decimal number as a string
#' @export
#' @examples
#' hexadecimal_to_decimal('ABC123')
hexadecimal_to_decimal <- function(hex){
  hex <- as.character(hex)
  if(length(hex) > 1){
    stop('hexadecimal_to_decimal function only accepts a single number')
  }
  hexVector <- stringr::str_split(hex,'',simplify = F)[[1]]
  if(length(hexVector) > 14 | (length(hexVector) == 14 & sum(as.numeric(hexVector[stringr::str_detect(hexVector,"[0-9]")]),na.rm = T) > 4)){
    stop(paste0(hex,' is too large for this package, sorry! (max = "40000000000000")'))
  }else{
    pow <- (length(hexVector)-1):0
    dec <- 0
    for(x in 1:length(hexVector)){
      dec <- dec + as.numeric((strtoi(hexVector[x],16)*(16**pow[x])))
    }
  }
  as.character(dec)
}


#' Binary to Decimal converter
#' @param bin A single binary number as a string - nchar(bin) <= 54
#' @return A single decimal number as a string
#' @export
#' @examples
#' binary_to_decimal('1001100')
binary_to_decimal <- function(bin){
  bin <- as.character(bin)
  if(length(bin) > 1){
    stop('binary_to_decimal function only accepts a single number')
  }
  if(stringr::str_count(bin,'1')+stringr::str_count(bin,'0') != nchar(bin)){
    stop(paste0(bin,' contains non-binary digits or characters'))
  }
  binVector <- stringr::str_split(bin,"",simplify = F)[[1]]
  # https://stackoverflow.com/questions/24009982/remove-zeros-in-the-start-and-end-of-a-vector
  if(min(which(binVector != 0)) == Inf){
    return('0')
  }else{
    binVector <- binVector[min(which(binVector != 0)):length(binVector)] # remove leading zero's
    if(length(binVector) > 55 | (length(binVector) == 55 & sum(as.numeric(binVector)) > 1)){
      stop(paste0(bin,' is too large for this package, sorry! (max = "1000000000000000000000000000000000000000000000000000000")'))
    }else{
      powerSeq <- seq(length(binVector)-1,0)
      dec <- 0
      for(i in 1:length(binVector)){
        dec <- dec + (strtoi(binVector[i]) * (2**powerSeq[i]))
      }
    }
    as.character(dec)
  }
}

#' Decimal to Binary converter
#' @param dec A single decimal number as a string
#' @return A single binary number as a string
#' @export
#' @examples
#' decimal_to_binary('1234')
decimal_to_binary <- function(dec){
  dec <- as.numeric(dec)
  bin <- c() # initialise empty vector to capture binary characters
  if(dec == 0){
    return('0000')
  }else{
    while(dec != 0){
      bin <- append(bin,dec%%2,after = 0)
      dec <- dec%/%2
    }
    bin <- paste0(bin,collapse = "")
    # always return whole nibbles
    if(nchar(bin)%%4){
      return(stringr::str_pad(bin,width = nchar(bin)+(4-nchar(bin)%%4),pad = "0",side = 'left'))
    }else{
      return(bin)
    }
  }
}

#' Decimal to Hexadecimal converter
#' @param dec A single decimal number as a string
#' @return A single hexadecimal number as a string
#' @export
#' @examples
#' decimal_to_hexadecimal('1234')
decimal_to_hexadecimal <- function(dec){
  dec <- as.numeric(dec)
  hex <- c() # initialise empty vector to capture hex characters
  if(dec == 0){
    return('0')
  }else{
    while(dec != 0){
      hex <- append(hex, as.character(as.hexmode(dec%%16)),after = 0)
      dec <- dec%/%16
    }
    return(paste0(hex,collapse = ""))
  }
}

#' Hexadecimal to Binary converter
#' This is simply a wrapper for decimal_to_binary and hexadecimal_to_decimal
#' @param hex A single hexadecimal number as a string
#' @return A single binary number as a string
#' @export
#' @examples
#' hexadecimal_to_binary('ABC123')
hexadecimal_to_binary <- function(hex){
  decimal_to_binary(hexadecimal_to_decimal(hex))
}

#' Binary to Hexadecimal converter
#' @param bin A single binary number as a string
#' @return A single hexadecimal number as a string
#' @export
#' @examples
#' binary_to_hexadecimal('1001101')
binary_to_hexadecimal <- function(bin){
  bin <- as.character(bin)
  if(as.numeric(bin) == 0){
    return('0')
  }else{
    if(nchar(bin)%%4){
      fullNibbleWidth <- nchar(bin)+(4-nchar(bin)%%4)
    }else{
      fullNibbleWidth <- nchar(bin)
    }
    binPadded <- stringr::str_pad(string = bin,
      width = fullNibbleWidth,
      side = 'left',
      pad = '0') # ensure binary string has whole nibbles
    binClustered <- stringi::stri_sub(binPadded,
      seq(1,nchar(binPadded),by=4),
      length=4) # convert to nibbles (4 bit chunks)
    dec <- strtoi(binClustered,2)
    hex <- as.hexmode(dec)
    paste0(hex,collapse = "")
  }

}


R/to_isodecimal_tform.R
#' ISO 11784 Dot Hexadecimal format To ISO11784 15 Digit Decimal
#' @param .data A vector of Dot Hexadecimal format strings.
#' @return A vector of ISO11784 15 Digit Decimal format strings.
#' @export
#' @examples
#' isodothex_to_isodecimal(c('3DD.3BC4567890'))
isodothex_to_isodecimal <- function(.data){

  formatTest <- ISO11784Tools::get_iso11784_format(as.character(.data)) == "isodothex"

  out <- as.character(.data)
  out[!formatTest] <- NA

  if(!all(is.na(out))){
    manufacturer <- out
    animalID <- out

    # extract manufacturer (left) and animal id (right) components
    manufacturer[formatTest] <- stringr::str_split(manufacturer[formatTest],pattern = '\\.',simplify = T)[,1]
    animalID[formatTest] <- stringr::str_split(animalID[formatTest],pattern = '\\.',simplify = T)[,2]

    # calculations LEFT
    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::hexadecimal_to_decimal)
    # calculations RIGHT
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::hexadecimal_to_decimal)
    # Leading zero's are removed in this process and need added back on, animalID only
    animalID[formatTest] <-  stringr::str_pad(animalID[formatTest], width = 12, side = 'left',pad = '0')

    out[formatTest] <- paste0(manufacturer[formatTest],animalID[formatTest])
  }

  out

}

#' ISO 11784 Raw Hexadecimal format, animal ID on the LEFT, To ISO11784 15 Digit Decimal
#' @param .data A vector of Raw Hexadecimal format strings.
#' @return A vector of ISO11784 15 Digit Decimal format strings.
#' @export
#' @examples
#' iso64bitleft_to_isodecimal(c('8000ABCDEF123456'))
iso64bitleft_to_isodecimal <- function(.data){

  formatTest <- ISO11784Tools::get_iso11784_format(as.character(.data)) == "iso64bitl"

  out <- as.character(.data)
  out[!formatTest] <- NA

  if(!all(is.na(out))){
    manufacturer <- out
    animalID <- out

    # extract manufacturer (left) and animal id (right) components
    # left transformations and calculations
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],5,7)
    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::hexadecimal_to_binary)
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],1,10)
    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::binary_to_decimal)
    #right transformations and calculations
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],7,16)
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::hexadecimal_to_binary)
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],3,40)
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::binary_to_decimal)
    # Leading zero's are removed in this process and need added back on, animalID only
    animalID[formatTest] <-  stringr::str_pad(animalID[formatTest],width = 12,pad = "0",side = 'left')

    out[formatTest] <- paste0(manufacturer[formatTest],animalID[formatTest])
  }

  out
}

#' ISO 11784 Raw Hexadecimal format, animal ID on the RIGHT, To ISO11784 15 Digit Decimal
#' @param .data A vector of Raw Hexadecimal format strings.
#' @return A vector of ISO11784 15 Digit Decimal format strings.
#' @export
#' @examples
#' iso64bitright_to_isodecimal(c('ABCDEF1234560001'))
iso64bitright_to_isodecimal <- function(.data){

  formatTest <- ISO11784Tools::get_iso11784_format(as.character(.data)) == "iso64bitr"

  out <- as.character(.data)
  out[!formatTest] <- NA

  if(!all(is.na(out))){
    manufacturer <- out
    animalID <- out

    # extract manufacturer (left) and animal id (right) components
    # left transformations and calculations
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],10,12)
    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::hexadecimal_to_binary)
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],3,12)
    manufacturer[formatTest] <- stringi::stri_reverse(manufacturer[formatTest])
    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::binary_to_decimal)
    #right transformations and calculations
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],1,10)
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::hexadecimal_to_binary)
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],1,38)
    animalID[formatTest] <- stringi::stri_reverse(animalID[formatTest])
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::binary_to_decimal)
    # Leading zero's are removed in this process and need added back on, animalID only
    animalID[formatTest] <-  stringr::str_pad(animalID[formatTest],width = 12,pad = "0",side = 'left')

    out[formatTest] <- paste0(manufacturer[formatTest],animalID[formatTest])
  }

  out
}


R/from_isodecimal_tform.R
#' ISO11784 15 Digit Decimal format To ISO11784 Dot Hexadecimal
#' @param .data A vector of ISO11784 15 Digit Decimal format strings.
#' @return A vector of Dot Hexadecimal format strings.
#' @export
#' @examples
#' isodecimal_to_isodothex(c('989737733408912'))
isodecimal_to_isodothex <- function(.data){

  formatTest <- ISO11784Tools::get_iso11784_format(as.character(.data)) == "isodecimal"

  out <- as.character(.data)
  out[!formatTest] <- NA

  if(!all(is.na(out))){
    manufacturer <- out
    animalID <- out

    # extract manufacturer (left) and animal id (right) components
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],1,3)
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],4,15)

    # calculations LEFT
    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::decimal_to_hexadecimal)
    # calculations RIGHT
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::decimal_to_hexadecimal)
    # Leading zero's are removed in this process and need added back on, animalID only
    animalID[formatTest] <-  stringr::str_pad(animalID[formatTest], width = 10, side = 'left',pad = '0')

    out[formatTest] <- stringr::str_to_upper(paste0(manufacturer[formatTest],'.',animalID[formatTest]))
  }

  out

}

#' ISO11784 15 Digit Decimal format To ISO11784 Raw Hexadecimal format, animal ID on the LEFT
#' @param .data A vector of ISO11784 15 Digit Decimal format strings.
#' @return A vector of Raw Hexadecimal format strings.
#' @export
#' @examples
#' isodecimal_to_iso64bitleft(c('989737733408912'))
isodecimal_to_iso64bitleft <- function(.data){

  formatTest <- ISO11784Tools::get_iso11784_format(as.character(.data)) == "isodecimal"

  out <- as.character(.data)
  out[!formatTest] <- NA

  if(!all(is.na(out))){
    manufacturer <- out
    animalID <- out

    # extract manufacturer (left) and animal id (right) components
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],1,3)
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],4,15)

    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::decimal_to_binary)
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::decimal_to_binary)

    # LEFT hand checks (manufacturer)
    # Force length 10 bits (assumption made that manafacturer <= 999)
    manufacturer[formatTest] <- stringr::str_trunc(manufacturer[formatTest],width = 10,side = 'left',ellipsis = '')
    manufacturer[formatTest] <- stringr::str_pad(manufacturer[formatTest],width = 10,pad = "0",side = 'left')

    # RIGHT hand checks (animalID)
    # Force length 10 bits (assumption made that animalID <= 274877906943)
    animalID[formatTest] <- stringr::str_trunc(animalID[formatTest],width = 38,side = 'left',ellipsis = '')
    animalID[formatTest] <- stringr::str_pad(animalID[formatTest],width = 38,pad = "0",side = 'left')

    out[formatTest] <- paste0('1000000000000000',manufacturer[formatTest],animalID[formatTest])

    out[formatTest] <- lapply(out[formatTest],ISO11784Tools::binary_to_hexadecimal)

  }

  stringr::str_to_upper(as.vector(out))
}

#' ISO11784 15 Digit Decimal format To ISO11784 Raw Hexadecimal format, animal ID on the RIGHT
#' @param .data A vector of ISO11784 15 Digit Decimal format strings.
#' @return A vector of Raw Hexadecimal format strings.
#' @export
#' @examples
#' isodecimal_to_iso64bitright(c('989737733408912'))
isodecimal_to_iso64bitright <- function(.data){

  formatTest <- ISO11784Tools::get_iso11784_format(as.character(.data)) == "isodecimal"

  out <- as.character(.data)
  out[!formatTest] <- NA

  if(!all(is.na(out))){
    manufacturer <- out
    animalID <- out

    # extract manufacturer (left) and animal id (right) components
    manufacturer[formatTest] <- stringr::str_sub(manufacturer[formatTest],1,3)
    animalID[formatTest] <- stringr::str_sub(animalID[formatTest],4,15)

    manufacturer[formatTest] <- lapply(manufacturer[formatTest],ISO11784Tools::decimal_to_binary)
    animalID[formatTest] <- lapply(animalID[formatTest],ISO11784Tools::decimal_to_binary)

    # LEFT hand checks (manufacturer)
    # Force length 10 bits (assumption made that manafacturer <= 999)
    manufacturer[formatTest] <- stringr::str_trunc(manufacturer[formatTest],width = 10,side = 'left',ellipsis = '')
    manufacturer[formatTest] <- stringr::str_pad(manufacturer[formatTest],width = 10,pad = "0",side = 'left')

    # RIGHT hand checks (animalID)
    # Force length 10 bits (assumption made that animalID <= 274877906943)
    animalID[formatTest] <- stringr::str_trunc(animalID[formatTest],width = 38,side = 'left',ellipsis = '')
    animalID[formatTest] <- stringr::str_pad(animalID[formatTest],width = 38,pad = "0",side = 'left')

    out[formatTest] <- paste0('1000000000000000',manufacturer[formatTest],animalID[formatTest])
    out[formatTest] <- stringi::stri_reverse(out[formatTest])
    out[formatTest] <- lapply(out[formatTest],ISO11784Tools::binary_to_hexadecimal)

  }

  stringr::str_to_upper(as.vector(out))
}


R/utilities.R
#' Convert a list of codes to all other formats
#' @param .data A vector of character strings
#' @return  A tibble containing the original codes with a detected format column plus a column for each known format
#' @export
#' @examples
#' convert_to_all(c('3DD.ABC4567890'))
convert_to_all <- function(.data){

  out <- tibble::as_tibble(as.character(.data))
  out$format <- get_iso11784_format(.data)
  out$isodecimal <- convert_to_isodecimal(.data)
  out$isodothex <- convert_to_isodothex(.data)
  out$iso64bitl <- convert_to_iso64bitl(.data)
  out$iso64bitr <- convert_to_iso64bitr(.data)

  out
}

#' ISO11784 Pattern detection routine to identify (as far as possible) the input format
#' @param .data A vector of character strings
#' @return  A vector of character strings, each item being one of 5 formats: c('unknown', 'isodecimal', 'isodothex','iso64bitl', 'iso64bitr')
#' @export
#' @examples
#' get_iso11784_format(c('3DD.ABC4567890'))
get_iso11784_format <- function(.data){

  ISOdothex <- "^[:xdigit:]{3}[\\.]{1}[:xdigit:]{10}$"
  ISO64bitl <- "^8000[:xdigit:]{12}$" # 64 bits and animal tag bit on the left
  ISO64bitr <- "^[:xdigit:]{12}0001$" # 64 bits and animal tag bit on the right
  ISOdecimal <- "^[0-9]{15}$"

  out <- tibble::as_tibble(as.character(.data)) #convert to tibble and ensure
  out[is.na(out)] <- "" # replace all NA values with empty string
  out$format <- "unknown"

  # capture basic format detection comparisons
  out[stringr::str_detect(out$value,ISOdecimal),]$format <- 'isodecimal'
  out[stringr::str_detect(out$value,ISOdothex),]$format <- 'isodothex'
  out[stringr::str_detect(out$value,ISO64bitl),]$format <- 'iso64bitl'
  out[stringr::str_detect(out$value,ISO64bitr),]$format <- 'iso64bitr'
  # reset outsider cases back to unknown
  out[out$format == 'isodecimal',][as.numeric(stringr::str_sub(out[out$format == 'isodecimal',]$value,4,-1)) > 274877906943,]$format <- 'unknown' # this number is the biggest 38 bit binary number, animal ID cannot be larger than 38 bits
  out[out$format == 'isodothex',][as.hexmode(stringr::str_sub(out[out$format == 'isodothex',]$value,5,5)) > as.hexmode('3'),]$format <- 'unknown' # this number is the biggest 38 bit binary number, animal ID cannot be larger than 38 bits
  out[out$format == 'iso64bitl',][!(stringr::str_detect(out[out$format == 'iso64bitl',]$value,"[a-fA-F]")),]$format <- 'unknown' # cautiously assume that 64bit hexadecimal should have at least 1 alpha character
  out[out$format == 'iso64bitr',][!(stringr::str_detect(out[out$format == 'iso64bitr',]$value,"[a-fA-F]")),]$format <- 'unknown' # cautiously assume that 64bit hexadecimal should have at least 1 alpha character

  out$format
}


#' Convert any recognised ISO11784/5 format to Decimal format
#' @param .data A vector of character strings
#' @return A converted vector of character strings
#' @export
#' @examples
#' convert_to_isodecimal(c('3E7.02DFDC1C35','8000F9C2DFDC1C36','EC383BFB439F0001'))
convert_to_isodecimal <- function(.data){

    ISOdecimal <- isodothex_to_isodecimal(isodecimal_to_isodothex(.data))
    ISOdothex <- isodothex_to_isodecimal(.data)
    ISO64bitLeft <- iso64bitleft_to_isodecimal(.data)
    ISO64bitRight <- iso64bitright_to_isodecimal(.data)
    return(dplyr::coalesce(ISOdecimal,ISOdothex,ISO64bitLeft,ISO64bitRight))

}


#' Convert any recognised ISO11784/5 format to Dot Hexadecimal format
#' @param .data A vector of character strings
#' @return A converted vector of character strings
#' @export
#' @examples
#' convert_to_isodothex(c('999012345678901','8000F9C2DFDC1C36','EC383BFB439F0001'))
convert_to_isodothex <- function(.data){

    ISOdecimal <- isodecimal_to_isodothex(.data)
    ISOdothex <- isodecimal_to_isodothex(isodothex_to_isodecimal(.data))
    ISO64bitLeft <- isodecimal_to_isodothex(iso64bitleft_to_isodecimal(.data))
    ISO64bitRight <- isodecimal_to_isodothex(iso64bitright_to_isodecimal(.data))
    return(dplyr::coalesce(ISOdecimal,ISOdothex,ISO64bitLeft,ISO64bitRight))
}


#' Convert any recognised ISO11784/5 format to Raw Hexadecimal format, animal tag bit on the Left
#' @param .data A vector of character strings
#' @return A converted vector of character strings
#' @export
#' @examples
#' convert_to_iso64bitl(c('999012345678901','3E7.02DFDC1C36','EC383BFB439F0001'))
convert_to_iso64bitl <- function(.data){

    ISOdecimal <- isodecimal_to_iso64bitleft(.data)
    ISOdothex <- isodecimal_to_iso64bitleft(isodothex_to_isodecimal(.data))
    ISO64bitLeft <- isodecimal_to_iso64bitleft(iso64bitleft_to_isodecimal(.data))
    ISO64bitRight <- isodecimal_to_iso64bitleft(iso64bitright_to_isodecimal(.data))
    return(dplyr::coalesce(ISOdecimal,ISOdothex,ISO64bitLeft,ISO64bitRight))
}


#' Convert any recognised ISO11784/5 format to Raw Hexadecimal format, animal tag bit on the Right
#' @param .data A vector of character strings
#' @return A converted vector of character strings
#' @export
#' @examples
#' convert_to_iso64bitr(c('999012345678901','8000F9C2DFDC1C36','3E7.02DFDC1C37'))
convert_to_iso64bitr <- function(.data){

    ISOdecimal <- isodecimal_to_iso64bitright(.data)
    ISOdothex <- isodecimal_to_iso64bitright(isodothex_to_isodecimal(.data))
    ISO64bitLeft <- isodecimal_to_iso64bitright(iso64bitleft_to_isodecimal(.data))
    ISO64bitRight <- isodecimal_to_iso64bitright(iso64bitright_to_isodecimal(.data))
    return(dplyr::coalesce(ISOdecimal,ISOdothex,ISO64bitLeft,ISO64bitRight))
}
